src/com/squareup/ideaplugin/dagger/PsiConsultantImpl.java

```diff
@@ -139,31 +143,70 @@ public static PsiClass checkForLazyOrProvider(PsiParameter psiParameter) {
       return wrapperClass;
     }
 
-    return getPsiClass(psiParameter, wrapperClass, psiParameterType);
+    return getPsiClass(wrapperClass, psiParameterType);
   }
 
-  private static PsiClass getPsiClass(PsiElement psiElement, PsiClass wrapperClass,
-      PsiType psiFieldType) {
+  private static PsiClass getPsiClass(PsiClass wrapperClass, PsiType psiFieldType) {
     PsiClassType psiClassType = (PsiClassType) psiFieldType;
-    Project project = psiElement.getProject();
-    JavaPsiFacade javaPsiFacade = JavaPsiFacade.getInstance(project);
-    GlobalSearchScope globalSearchScope = GlobalSearchScope.allScope(project);
-
-    PsiClass lazyClass = javaPsiFacade.findClass(CLASS_LAZY, globalSearchScope);
-    PsiClass providerClass = javaPsiFacade.findClass(CLASS_PROVIDER, globalSearchScope);
 
     PsiClassType.ClassResolveResult classResolveResult = psiClassType.resolveGenerics();
     PsiClass outerClass = classResolveResult.getElement();
 
     // If Lazy<Foo> or Provider<Foo>, extract Foo as the interesting type.
-    if (outerClass != null //
-        && (outerClass.equals(lazyClass) || outerClass.equals(providerClass))) {
-      PsiType genericType =
-          classResolveResult.getSubstitutor().getSubstitutionMap().values().iterator().next();
+    if (PsiConsultantImpl.isLazyOrProvider(outerClass)) {
+      PsiType genericType = extractFirstTypeParameter(psiClassType);
       // Convert genericType to its PsiClass and store in psiClass
       wrapperClass = getClass(genericType);
     }
 
     return wrapperClass;
   }
+
+  public static boolean hasTypeParameters(PsiElement psiElement, List<PsiType> typeParameters) {
+    List<PsiType> actualTypeParameters = getTypeParameters(psiElement);
+    return actualTypeParameters.equals(typeParameters);
+  }
+
+  public static List<PsiType> getTypeParameters(PsiElement psiElement) {
```

Comment:
Why a List instead of a Set? I don't think ordering matters when it comes to type parameters.